File: Method/Statistical_methods/Frequency_domain/sd.py
"""
Scalogram-based Discrimination (SD) method for pulse shape discrimination.

This module implements a wavelet-based approach to pulse shape discrimination
using continuous wavelet transform (CWT) analysis. The method generates a
discrimination mask from labeled training data and uses it to compute
discrimination factors for mixed signals.

Reference:
- Abdelhakim, Assem, and Ehab Elshazly.
  "Efficient pulse shape discrimination using scalogram image masking and decision tree."
  Nuclear Instruments and Methods in Physics Research Section A: Accelerators, Spectrometers,
  Detectors and Associated Equipment 1050 (2023): 168140.
"""

import numpy as np
import pywt
from tqdm import tqdm
from typing import Optional


def _normalize_signals(signals: np.ndarray) -> np.ndarray:
    """
    Normalize each pulse signal by its maximum absolute value.

    Args:
        signals: Input signals array of shape (n_signals, signal_length)

    Returns:
        Normalized signals array of the same shape
    """
    normalized = np.zeros_like(signals)
    for idx in range(signals.shape[0]):
        pulse = signals[idx, :]
        max_amplitude = np.max(np.abs(pulse))
        if max_amplitude < 1e-9:  # Avoid division by near-zero values
            normalized[idx, :] = pulse
        else:
            normalized[idx, :] = pulse / max_amplitude
    return normalized


def _convert_to_grayscale(matrix: np.ndarray) -> np.ndarray:
    """
    Convert a matrix to grayscale values (0-255).

    Args:
        matrix: Input matrix to be converted to grayscale

    Returns:
        Grayscale matrix with values between 0 and 255
    """
    min_val, max_val = matrix.min(), matrix.max()
    return 255 * (matrix - min_val) / (max_val - min_val + 1e-9)


def _generate_discrimination_mask(
        class1_signals: np.ndarray,
        class2_signals: np.ndarray,
        wavelet: str = 'mexh',
        scales: Optional[np.ndarray] = None,
        threshold: int = 127
) -> np.ndarray:
    """
    Generate a binary discrimination mask from training data using wavelet analysis.

    Args:
        class1_signals: Training data for first class of signals
        class2_signals: Training data for second class of signals
        wavelet: Wavelet type to use (default: 'mexh')
        scales: Wavelet scales (default: np.arange(1,51))
        threshold: Grayscale threshold for binary conversion (default: 127)

    Returns:
        Binary mask for signal discrimination

    Raises:
        ValueError: If no discrimination features are found between signal classes
    """
    if scales is None:
        scales = np.arange(1, 51)

    # Use a subset of training samples to build the mask
    num_samples = min(class1_signals.shape[0], class2_signals.shape[0])
    training_size = max(1, int(num_samples / 15))
    difference_count = np.zeros((len(scales), class1_signals.shape[1]), dtype=int)

    for sample_idx in range(training_size):
        # Process one pulse from each class
        pulse1 = class1_signals[sample_idx, :]
        pulse2 = class2_signals[sample_idx, :]

        # Calculate CWT coefficients and corresponding energy scalograms
        coeffs1, _ = pywt.cwt(pulse1, scales, wavelet, method='fft')
        coeffs2, _ = pywt.cwt(pulse2, scales, wavelet, method='fft')
        energy1 = np.abs(coeffs1) ** 2
        energy2 = np.abs(coeffs2) ** 2

        # Convert energy scalograms to grayscale and then to binary ROI images
        gray1 = _convert_to_grayscale(energy1)
        gray2 = _convert_to_grayscale(energy2)
        roi1 = np.where(gray1 >= threshold, 1, 0)
        roi2 = np.where(gray2 >= threshold, 1, 0)

        # Accumulate absolute differences between the ROIs
        difference_count += np.abs(roi1 - roi2)

    nonzero_count = np.count_nonzero(difference_count)
    if nonzero_count == 0:
        raise ValueError("No discrimination features found between signal classes")

    difference_threshold = difference_count.sum() / nonzero_count
    mask = np.where(difference_count >= difference_threshold, 1, 0)
    return mask


def _discriminate_signals(
        mixed_signals: np.ndarray,
        discrimination_mask: np.ndarray,
        wavelet: str = 'mexh',
        scales: Optional[np.ndarray] = None,
        threshold: int = 127
) -> np.ndarray:
    """
    Compute discrimination factors for each signal using the provided mask.

    Args:
        mixed_signals: Array of mixed signals
        discrimination_mask: Pre-generated discrimination mask
        wavelet: Wavelet type (default: 'mexh')
        scales: Wavelet scales (default: np.arange(1,51))
        threshold: Grayscale threshold (default: 127)

    Returns:
        Discrimination factors for each signal
    """
    if scales is None:
        scales = np.arange(1, 51)

    num_signals = mixed_signals.shape[0]
    factors = np.zeros(num_signals)

    for idx in tqdm(range(num_signals), desc="Calculating Discrimination Factors", unit="signal"):
        signal = mixed_signals[idx, :]
        coeffs, _ = pywt.cwt(signal, scales, wavelet, method='fft')
        scalogram = np.abs(coeffs) ** 2
        gray_scalogram = _convert_to_grayscale(scalogram)
        roi = np.where(gray_scalogram >= threshold, 1, 0)
        # Apply the discrimination mask and compute the factor
        masked_roi = roi[discrimination_mask == 1]
        factors[idx] = np.sum(masked_roi) / np.sum(discrimination_mask)
    return factors


def get_psd_factor(pulse_signal: np.ndarray) -> np.ndarray:
    """
    Compute pulse shape discrimination factors using the Scalogram-based
    Discrimination (SD) method.

    The function performs the following steps:
      - Loads labeled training data for class 1 and class 2 signals
      - Normalizes all signals by their maximum absolute values
      - Generates a discrimination mask using wavelet analysis
      - Applies the mask to compute discrimination factors

    Args:
        pulse_signal: Array of mixed pulse signals

    Returns:
        Array of discrimination factors for the input signals
    """
    # SD requires labeled signals to generate mask
    class1_signals = np.loadtxt('Data/Train/EJ299_33_AmBe_9414_gamma_train.txt')
    class2_signals = np.loadtxt('Data/Train/EJ299_33_AmBe_9414_neutron_train.txt')

    # Normalize the signals
    class1_signals = _normalize_signals(class1_signals)
    class2_signals = _normalize_signals(class2_signals)

    # Ensure the mixed signals are normalized
    mixed_signals = _normalize_signals(pulse_signal)

    # Generate discrimination mask
    mask = _generate_discrimination_mask(
        class1_signals,
        class2_signals,
        wavelet='mexh',
        scales=np.arange(1, 51),
        threshold=127
    )

    # Compute and return the discrimination factors for the mixed signals
    return _discriminate_signals(
        mixed_signals,
        mask,
        wavelet='mexh',
        scales=np.arange(1, 51),
        threshold=127
    )

File: MATLAB/Method/Statistical_methods/Frequency_domain/SD/SD.m
function [Pulse_shape_discrimination_factor] = SD(pulse_signal)

%--------------------------------------------------------------------------
% Performs pulse shape discrimination using Scalogram-based Discrimination (SD).
% 
% Uses continuous wavelet transform (CWT) to generate scalograms and applies
% a trained discrimination mask to classify mixed neutron/gamma signals.
% The mask is created from labeled training data of both signal types.
% 
% Reference:
% Abdelhakim, Assem, and Ehab Elshazly. "Efficient pulse shape discrimination
% using scalogram image masking and decision tree." Nuclear Instruments and
% Methods in Physics Research Section A: Accelerators, Spectrometers, 
% Detectors and Associated Equipment 1050 (2023): 168140.
%--------------------------------------------------------------------------
    % Load and normalize reference signals
    class1_signals = load('Data/Reference_signal/EJ299_33_AmBe_9414_neutron_ref.txt');
    class2_signals = load('Data/Reference_signal/EJ299_33_AmBe_9414_gamma_ref.txt');
    class1_signals = normalize_signals(class1_signals);
    class2_signals = normalize_signals(class2_signals);

    % Process input signals
    mixed_signals = normalize_signals(pulse_signal);

    % Create discrimination mask from training data
    mask = generate_discrimination_mask(class1_signals, class2_signals, 'mexh', 1:50, 127);

    % Apply mask to get discrimination factors
    Pulse_shape_discrimination_factor = discriminate_signals(mixed_signals, mask, 'mexh', 1:50, 127);
end

function normalized_signals = normalize_signals(signals)
    % Normalize signals by their maximum absolute values
    max_abs_value = max(abs(signals), [], 2);  % Row-wise maximum
    normalized_signals = signals ./ max_abs_value;
end

function grayscale_matrix = convert_to_grayscale(matrix)
    % Convert matrix values to grayscale range [0, 255]
    % 
    % Args:
    %   matrix: Input data matrix
    % Returns:
    %   Normalized matrix scaled to [0, 255]
    
    min_val = min(matrix(:));
    max_val = max(matrix(:));
    grayscale_matrix = 255 * (matrix - min_val) / (max_val - min_val + 1e-9);
end

function mask = generate_discrimination_mask(class1_signals, class2_signals, wavelet, scales, threshold)
    % Generate binary mask highlighting regions of maximum discrimination
    % between two signal classes using CWT scalograms
    
    % Default parameters
    if nargin < 3
        wavelet = 'mexh';     % Mexican hat wavelet
    end
    if nargin < 4
        scales = 1:50;        % Wavelet scales
    end
    if nargin < 5
        threshold = 127;      % Grayscale threshold
    end

    % Select training subset
    num_samples = min(size(class1_signals, 1), size(class2_signals, 1));
    training_size = max(1, floor(num_samples / 15));
    difference_count = zeros(length(scales), size(class1_signals, 2));

    % Accumulate differences between class scalograms
    for sample_idx = 1:training_size
        % Get sample pair
        pulse1 = class1_signals(sample_idx, :);
        pulse2 = class2_signals(sample_idx, :);

        % Calculate scalograms
        [coeffs1, ~] = cwt(pulse1, scales, wavelet, 'scal');
        [coeffs2, ~] = cwt(pulse2, scales, wavelet, 'scal');
        energy1 = abs(coeffs1) .^ 2;
        energy2 = abs(coeffs2) .^ 2;

        % Convert to binary ROIs
        gray1 = convert_to_grayscale(energy1);
        gray2 = convert_to_grayscale(energy2);
        roi1 = gray1 >= threshold;
        roi2 = gray2 >= threshold;

        % Add to difference map
        difference_count = difference_count + abs(roi1 - roi2);
    end

    % Validate and create mask
    nonzero_count = nnz(difference_count);
    if nonzero_count == 0
        error('No discrimination features found between signal classes');
    end

    % Create mask using average difference as threshold
    difference_threshold = sum(difference_count(:)) / nonzero_count;
    mask = difference_count >= difference_threshold;
end

function Pulse_shape_discrimination_factor = discriminate_signals(mixed_signals, mask, wavelet, scales, threshold)
    % Apply discrimination mask to mixed signals and compute discrimination factors
    
    % Default parameters
    if nargin < 4
        scales = 1:50;
    end
    if nargin < 5
        threshold = 127;
    end
    
    num_signals = size(mixed_signals, 1);
    Pulse_shape_discrimination_factor = zeros(num_signals, 1);
    
    % Process each signal
    for i = 1:num_signals
        signal = mixed_signals(i, :);
        [coeffs, ~] = cwt(signal, scales, wavelet, 'scal');
        scalogram = abs(coeffs).^2;
        gray_scalogram = convert_to_grayscale(scalogram);
        roi = gray_scalogram >= threshold;
        
        % Calculate discrimination factor as ratio of matching mask points
        masked_roi = roi(mask);
        Pulse_shape_discrimination_factor(i) = sum(masked_roi(:)) / sum(mask(:));
    end
end
